miio/cooker_multi.py

```diff
@@ -0,0 +1,419 @@
+
+        max_minutes = (self.profile_bytes[10] * 60) + self.profile_bytes[11]
+        min_minutes = (self.profile_bytes[12] * 60) + self.profile_bytes[13]
+
+        if minutes < min_minutes or minutes > max_minutes:
+            return
+
+        self.profile_bytes[8] = math.floor(minutes / 60)
+        self.profile_bytes[9] = minutes % 60
+
+        self.update_checksum()
+
+    def is_schedule_enabled(self):
+        return (self.profile_bytes[14] & 0x80) == 0x80
+
+    def set_schedule_enabled(self, enabled):
+        if enabled:
+            self.profile_bytes[14] |= 0x80
+        else:
+            self.profile_bytes[14] &= 0x7F
+
+        self.update_checksum()
+
+    def set_schedule_duration(self, duration):
+        """Set the schedule time (delay before cooking) in minutes."""
+        schedule_flag = self.profile_bytes[14] & 0x80
+        self.profile_bytes[14] = math.floor(duration / 60) & 0xFF
+        self.profile_bytes[14] |= schedule_flag
+        self.profile_bytes[15] = (duration % 60 | self.profile_bytes[15] & 0x80) & 0xFF
+
+        self.update_checksum()
+
+    def is_akw_enabled(self):
+        return (self.profile_bytes[15] & 0x80) == 0x80
+
+    def set_akw_enabled(self, enabled):
+        if enabled:
+            self.profile_bytes[15] |= 0x80
+        else:
+            self.profile_bytes[15] &= 0x7F
+
+        self.update_checksum()
+
+    def calc_checksum(self):
+        crc = crcmod.mkCrcFun(0x11021, rev=False, initCrc=0x0, xorOut=0x0)(
+            self.profile_bytes
+        )
+        checksum = bytearray(2)
+        checksum[0] = (crc >> 8) & 0xFF
+        checksum[1] = crc & 0xFF
+        return checksum
+
+    def update_checksum(self):
+        self.checksum = self.calc_checksum()
+
+    def is_valid(self):
+        return len(self.profile_bytes) == 174 and self.checksum == self.calc_checksum()
+
+    def get_profile_hex(self):
+        return (self.profile_bytes + self.checksum).hex()
+
+
+class CookerStatus(DeviceStatus):
+    def __init__(self, data):
+        self.data = data
+
+    @property
+    def mode(self) -> OperationMode:
+        """Current operation mode."""
+        return OperationMode(self.data["status"])
+
+    @property
+    def menu(self) -> str:
+        """Selected menu id."""
+        try:
+            return COOKING_MENUS[self.data["menu"]]
+        except KeyError:
+            return "Unknown menu"
+
+    @property
+    def stage(self) -> str:
+        """Current stage if cooking."""
+        try:
+            return COOKING_STAGES[self.data["phase"]]["name"]
+        except KeyError:
+            return "Unknown stage"
+
+    @property
+    def temperature(self) -> Optional[int]:
+        """Current temperature, if idle.
+
+        Example values: 29
+        """
+        return self.data["temp"]
+
+    @property
+    def start_time(self) -> int:
+        """Start time of cooking?"""
+        return int(self.data["t_start"])
+
+    @property
+    def remaining(self) -> int:
```

Comment:
And this should also likely be `ready_at(self) -> datetime` or similar to avoid spamming the homeassistant state machine for updates.

Reply:
Yes, agree. I change it to calculate a datetime from t_left and t_pre. I also cut millis and seconds from the result to prevent fluctuations.

Reply:
Great, thanks :+1:

Reply:
I found a problem about this. Usually the cooking process is a PreCook phase which is a scheduled delay before the actual Running phase starts. During the PreCook phase the cooker only reports the remaining minutes left. If I compute a timestamp here and the pull rate from home assistant is higher than 1 minute the resulting timestamp will fluctuate. For the Running phase the problem is not that bad because in this phase the cooker is providing remaining seconds. Right now I left the "remaining" property which provides remaining minutes and the "ready_at" property for reference.

tl;dr:
"remaining" property will spam HA with the remaining minutes; "ready_at" property will spam HA with a fluctuating timestamp (03:14 -> 03:15 -> 03:14 -> 03:15 ...)

Any idea for a clean solution?

Reply:
I removed the ready_at sensor in favor of the remaining sensor to keep the multicooker integration similar to the other cooker integration

Reply:
Ah, perhaps some sort of rounding could have worked here. Anyway, as it's now removed I'll just mark this as resolved.